### ISSUE TITLE
DllNotFoundException if Library deleted before open Unity

### ISSUE THREAD

**Issue by IvanMurzak** [firebase/quickstart-unity#587]
Unity editor version: **2019.2.21** (the same on **2019.1.14**)
Firebase Unity SDK version: **6.10.0** (the same on **6.6.0**)
Firebase plugins in use (Auth, Database, etc.): **Analytics, Messages, RemoteConfigs, Crashlytics**
Additional SDKs you are using (Facebook, AdMob, etc.): **Appodeal**
Platform you are using the Unity editor on (Mac, Windows, or Linux): **Windows**
Platform you are targeting (iOS, Android, and/or desktop): **Android**
Scripting Runtime (Mono, and/or IL2CPP): **IL2CPP (.net 4)**

### **Info**

After implemented CI integration with Gitlab and made auto-build machine in unity batchmode and removing "Library" before executing build command I got DllNotFoundException

### **Command**

`call %UNITY_EXE_PATH% %APK% -projectPath .\ -buildTarget Android -logFile build.log -executeMethod BuildProject.BuildAndroid -quit -batchmode -nographics `

### **BuildProject.cs**

```
using System;
using System.Linq;
using UnityEditor;

public static class BuildProject
{
	static string[] GetScenePaths() => EditorBuildSettings.scenes.Select(x => x.path).ToArray();

    /**
     * returns false if one or more required environment variables are not defined
     * */
    static bool EnvironmentVariablesMissing(string[] envvars)
    {
        string value;
        var missing = false;
        foreach (var envvar in envvars)
        {
            value = Environment.GetEnvironmentVariable(envvar);
            if (value == null)
            {
                Console.Write("BUILD ERROR: Required Environment Variable is not set: ");
                Console.WriteLine(envvar);
                missing = true;
            }
        }

        return missing;
    }

	/**
     * Main entry point
     * - check if all required environment variables are defined
     * - configure the android build
     * - build the apk (path read from the command line argument)
     */
	public static void BuildAndroid()
    {
        var envvars = new string[]
        {
			"ANDROID_KEYSTORE_NAME",
			"ANDROID_KEYSTORE_PASSWORD",
			"ANDROID_KEYALIAS_NAME",
			"ANDROID_KEYALIAS_PASSWORD",
			"ANDROID_SDK_ROOT"
        };
        if (EnvironmentVariablesMissing(envvars))
        {
            Environment.ExitCode = -1;
            return; // note, we can not use Environment.Exit(-1) - the buildprocess will just hang afterwards
        }

		GooglePlayServices.PlayServicesResolver.Resolve(forceResolution: true);
		DebugFormat.Log<string>($"PlayServicesResolver force resolved");
		AssetDatabase.ImportAsset("Assets/!PROJECT/Scripts/Project.Scripts.asmdef"); // trying to fix DLLNotFoundException
		DebugFormat.Log<string>($"Reimporting Project.Scripts.asmdef completed");
		// Available Playersettings: https://docs.unity3d.com/ScriptReference/PlayerSettings.Android.html

		//set the internal apk version to the current unix timestamp, so this increases with every build
		PlayerSettings.Android.bundleVersionCode = (Int32)(DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalSeconds; 

        //set the other settings from environment variables
        PlayerSettings.Android.keystoreName = Environment.GetEnvironmentVariable("ANDROID_KEYSTORE_NAME");
        PlayerSettings.Android.keystorePass = Environment.GetEnvironmentVariable("ANDROID_KEYSTORE_PASSWORD");
        PlayerSettings.Android.keyaliasName = Environment.GetEnvironmentVariable("ANDROID_KEYALIAS_NAME");
        PlayerSettings.Android.keyaliasPass = Environment.GetEnvironmentVariable("ANDROID_KEYALIAS_PASSWORD");

		EditorPrefs.SetString("AndroidSdkRoot",		Environment.GetEnvironmentVariable("ANDROID_SDK_ROOT"));
		EditorPrefs.GetString("AndroidNdkRootR16b",	Environment.GetEnvironmentVariable("ANDROID_SDK_NDK_R16")); // G:/android-sdk/android-ndk-r16b
		EditorPrefs.GetString("AndroidNdkRoot",		Environment.GetEnvironmentVariable("ANDROID_SDK_NDK"));		// G:/android-sdk/android-ndk-r13b

		//Get the apk file to be built from the command line argument
		var outputapk = Environment.GetCommandLineArgs().Skip(1).First();
		DebugFormat.Log<string>($"APK output = {outputapk}");
		BuildPipeline.BuildPlayer(GetScenePaths(), outputapk, BuildTarget.Android, BuildOptions.None);
    }
}
```

### **DllNotFoundException**

```
Uploading Crash Report
DllNotFoundException: FirebaseCppApp-6_10_0
  at (wrapper managed-to-native) Firebase.AppUtilPINVOKE+SWIGExceptionHelper.SWIGRegisterExceptionCallbacks_AppUtil(Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate,Firebase.AppUtilPINVOKE/SWIGExceptionHelper/ExceptionDelegate)
  at Firebase.AppUtilPINVOKE+SWIGExceptionHelper..cctor () [0x000ee] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\AppUtilPINVOKE.cs:117 
Rethrow as TypeInitializationException: The type initializer for 'SWIGExceptionHelper' threw an exception.
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_generic_class_init(intptr)
  at Firebase.AppUtilPINVOKE..cctor () [0x00000] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\AppUtilPINVOKE.cs:137 
Rethrow as TypeInitializationException: The type initializer for 'Firebase.AppUtilPINVOKE' threw an exception.
  at Firebase.AppUtil.SetLogFunction (Firebase.LogUtil+LogMessageDelegate arg0) [0x00000] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\AppUtil.cs:64 
  at Firebase.LogUtil..ctor () [0x00006] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\LogUtil.cs:80 
  at Firebase.LogUtil..cctor () [0x00010] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\LogUtil.cs:36 
Rethrow as TypeInitializationException: The type initializer for 'Firebase.LogUtil' threw an exception.
  at Firebase.FirebaseApp..cctor () [0x0004a] in Z:\tmp\tmp.5kTaS2UUGh\firebase\app\client\unity\proxy\FirebaseApp.cs:62 
Rethrow as TypeInitializationException: The type initializer for 'Firebase.FirebaseApp' threw an exception.
  at Firebase.Crashlytics.Crashlytics+PlatformAccessor..cctor () [0x00014] in Z:\tmp\tmp.c5tnl4BIX3\firebase\crashlytics\client\unity\proxy\Crashlytics.cs:101 
Rethrow as TypeInitializationException: The type initializer for 'PlatformAccessor' threw an exception.
  at Firebase.Crashlytics.Crashlytics.Log (System.String message) [0x00000] in Z:\tmp\tmp.c5tnl4BIX3\firebase\crashlytics\client\unity\proxy\Crashlytics.cs:49 
  at DebugFormat.Log[T] (T instance, System.String message) [0x00017] in F:\Programs\GitLab-Runner\builds\1psxyZNY\0\IvanMurzak\TD-World\Assets\!PROJECT\Extensions\Unity Extensions\Utils\DebugFormat.cs:41 
UnityEngine.DebugLogHandler:Internal_LogException(Exception, Object)
UnityEngine.DebugLogHandler:LogException(Exception, Object)
UnityEngine.Logger:LogException(Exception, Object)
UnityEngine.Debug:LogException(Exception)
DebugFormat:Log(WalletSO, String) (at Assets\!PROJECT\Extensions\Unity Extensions\Utils\DebugFormat.cs:41)
WalletSO:OnDisable() (at Assets\!PROJECT\Scripts\Data\WalletSO.cs:57)
UnityEditor.BuildPipeline:BuildPlayerInternalNoCheck(String[], String, String, BuildTargetGroup, BuildTarget, BuildOptions, Boolean)
UnityEditor.BuildPipeline:BuildPlayerInternal(String[], String, String, BuildTargetGroup, BuildTarget, BuildOptions) (at C:\buildslave\unity\build\Editor\Mono\BuildPipeline.bindings.cs:368)
UnityEditor.BuildPipeline:BuildPlayer(String[], String, String, BuildTargetGroup, BuildTarget, BuildOptions) (at C:\buildslave\unity\build\Editor\Mono\BuildPipeline.bindings.cs:267)
UnityEditor.BuildPipeline:BuildPlayer(BuildPlayerOptions) (at C:\buildslave\unity\build\Editor\Mono\BuildPipeline.bindings.cs:250)
UnityEditor.BuildPipeline:BuildPlayer(String[], String, BuildTarget, BuildOptions) (at C:\buildslave\unity\build\Editor\Mono\BuildPipeline.bindings.cs:245)
BuildProject:BuildAndroid() (at Assets\!PROJECT\Editor\BuildProject.cs:74)
 
(Filename: Assets/!PROJECT/Extensions/Unity Extensions/Utils/DebugFormat.cs Line: 41)
```

### **Additional info**
Tried with Unity 2019.1.14 and Firebase 6.6.0
and Unity 2019.2.21 and Firebase 6.6.0
and Unity 2019.2.21 and Firebase 6.10.0

In any combination, windows crashes with kernel error (not event blue screen) the PC just shutdown 
![image](https://user-images.githubusercontent.com/9135028/74764264-10db6680-528a-11ea-81aa-d1e6778905a5.png)
(Feb 18 are related crashes)

Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)? 
**Yes, every single time after push (~20 times in general)**

What's the issue repro rate? (eg 100%, 1/5 etc)
**100%**

**Comment by DellaBitta** [comment]
Hi @IvanMurzak,

This seems similar to [Issue 567](https://github.com/firebase/quickstart-unity/issues/567). Could you review it to see if it's the same problem?  Thanks!

**Comment by IvanMurzak** [comment]
@DellaBitta, hello! Sounds different, because in my case I can open Unity Editor and reimport any C# file it fixes the problem. 

The core of my problem is DllNotFound if execute Unity in batchmode (and -nographics -quit) without Library folder. I tried to reimport asmdef file for recompiling in BuildProject.cs, but it doesn't help. And batchmode compilation still fails. That's why I guess it is another problem.

**Comment by IvanMurzak** [comment]
Researched Windows crash. The reason is GitLab runner. You may just skip information about windows crash. It is not related to DllNotFoundException while compilation through batchmode when Library was removed

**Comment by DellaBitta** [comment]
Ok great, thanks, good to know!

**Comment by patm1987** [comment]
Hi @IvanMurzak ,

Would it be possible to share more of the log file? Perhaps there's some other error that was overlooked before?

I'm also curious if you can build on this machine outside of batch mode. Ex: just checkout your project, then open it in Unity itself.

Thanks,
--Patrick

**Comment by stewartmiles** [comment]
@IvanMurzak the stack trace contains Crashlytics in the stack...

```
Rethrow as TypeInitializationException: The type initializer for 'PlatformAccessor' threw an exception.
  at Firebase.Crashlytics.Crashlytics.Log (System.String message) [0x00000] in Z:\tmp\tmp.c5tnl4BIX3\firebase\crashlytics\client\unity\proxy\Crashlytics.cs:49 
  at DebugFormat.Log[T] (T instance, System.String message) [0x00017] in F:\Programs\GitLab-Runner\builds\1psxyZNY\0\IvanMurzak\TD-World\Assets\!PROJECT\Extensions\Unity Extensions\Utils\DebugFormat.cs:41 
```

While we support a chunk of Firebase features in desktop builds and in the Unity Editor context, we do not support Crashlytics so I think it's possible that's part of the problem.  Are you using Firebase in Unity editor scripts?  If so, that would explain why Firebase Crashlytics is being initialized adding a hook for C# crash reporting and then blowing up when it tries to log messages via the Firebase desktop shared library / DLL `DllNotFoundException: FirebaseCppApp-6_10_0` which sounds like it's missing in your project.  A shared library should be present for each desktop platform under `Assets/Firebase/Plugins/x86_64`, if `FirebaseCppApp*.(dll|bundle|so)` is missing then of since this is trying to call into the C layer of the plugin this will blow up.

So in summary, search your codebase for places where Firebase could be initialized in an editor script.  You can turn on debug logging using [Firebase.FirebaseApp.LogLevel](https://firebase.google.com/docs/reference/unity/class/firebase/firebase-app#class_firebase_1_1_firebase_app_1aa95be32e8c8d56783fcb16234fbe2b63) which _may_ help you track this down.
